### PR TITLE
Implement host_rewrite_path option

### LINKED ISSUE #12430: Proposal: Introduce host_rewrite_path support in route config
**Rational**

Using subdomains for tenant-id in a multi-tenant application is a relatively popular strategy, our own MediaWiki (the software behind Wikipedia) uses `Host` header to distinguish between different language wikis. Ability to rewrite `Host` based on path would provide flexibility in moving the tenant-id from subdomain to path when deploying existing software behind the proxy.

**Proposal**

Introduce a new `host_rewrite_path` option in `RouteAction:host_rewrite_specifier`. 

```
// Indicates that during forwarding, the host header will be swapped with the result of the
// given regex substitution run agains the request path. 
type.matcher.v3.RegexMatchAndSubstitute host_rewrite_path = 36;
```

The behavior of the new property would be the following: `Host` header will be set to the result of the `RegexMatchAndSubstitute`, executed agains the path. However, this would need to be applied with care, since the regex matcher needs to be guaranteed to match the whole path string. Alternatively, a new `RegexMatchAndReplace` message could introduced, where the result would replace the input string with capture groups applied, or null if the regex did not match the input - in that case the host header would not be swapped if the regex did not match.

We are going to provide implementation and testing. Seeking approval from the maintainers and guidance on whether to reuse `RegexMatchAndSubstitute` or introduce a new message type.

### PR DESCRIPTION
**Pull request by Pchelolo**
This implements a `host_rewrite_path` option for rewriting the Host header based on path. See rational in the linked issue.

Note: the regex is executed on the path with query/fragment stripped. This is analogues to what `regex_rewrite` option is doing.

Commit Message: Implement host_rewrite_path option
Risk Level: Low
Testing: added unit tests
Docs Changes: document the new option in proto file
Release Notes: added to current.rst
Fixes #12430 

Signed-off-by: Petr Pchelko <ppchelko@wikimedia.org>

### CODE REVIEW

File: api/envoy/config/route/v3/route_components.proto [hunk 1]
```diff
@@ -890,6 +890,11 @@ message RouteAction {
     //   must come from trusted source.
     string host_rewrite_header = 29
         [(validate.rules).string = {well_known_regex: HTTP_HEADER_NAME strict: false}];
+
+    // Indicates that during forwarding, the host header will be swapped with
+    // the result of the regex substitution executed on path value with query and fragment removed.
+    // This is useful for transitioning variable content between path segment and subdomain.
```
Comment (dio) [inline api/envoy/config/route/v3/route_components.proto:896]: While it is clear from the test cases, I think giving an example here will be useful.

File: source/common/router/config_impl.cc [hunk 2]
```diff
@@ -527,6 +534,10 @@ void RouteEntryImplBase::finalizeRequestHeaders(Http::RequestHeaderMap& headers,
         headers.setHost(header_value);
       }
     }
+  } else if (host_rewrite_path_ != nullptr) {
+    std::string path(headers.getPathValue());
```
Comment (dio) [inline source/common/router/config_impl.cc:538]: const?

**Comment by Pchelolo** [comment]
@dio thank you for the review. Addressed your comments.

File: source/common/router/config_impl.cc [hunk 3]
```diff
@@ -287,6 +287,13 @@ RouteEntryImplBase::RouteEntryImplBase(const VirtualHostImpl& vhost,
                                     ? absl::optional<Http::LowerCaseString>(Http::LowerCaseString(
                                           route.route().host_rewrite_header()))
                                     : absl::nullopt),
+      host_rewrite_path_(
```
Comment (asraa) [inline source/common/router/config_impl.cc:290]: nit: suggest using a "regex" in the naming so it's clear this is a regex

Comment (Pchelolo) [reply source/common/router/config_impl.cc:290]: IMHO that would make it less clear what the regex is being executed on. It seems like other options names suggest where the Host header rewrite is coming from, (e.g. `host_rewrite_literal`, `host_rewrite_header`) not how the result is constructed. 

With `host_rewrite_regex` you could imaging the regex being exectuted on some other header, or on the original Host name.

Having said that, happy to change if you think regex's better.

Comment (Pchelolo) [reply source/common/router/config_impl.cc:290]: Or, you mean `host_rewrite_path_regex`, not `host_rewrite_regex`?

Comment (asraa) [reply source/common/router/config_impl.cc:290]: Haha yep, definitely agree that path should be included so it's clear. On my own first glance though the naming didn't line up to my understanding that this was a compiled regex matcher. But no worries either way (header file's got the type anyway)

Comment (Pchelolo) [reply source/common/router/config_impl.cc:290]: ok. changing to `host_rewrite_path_regex`

**Comment by Pchelolo** [comment]
Merged master in to resolve a conflict.

File: source/common/router/config_impl.cc [hunk 4]
```diff
@@ -287,6 +287,14 @@ RouteEntryImplBase::RouteEntryImplBase(const VirtualHostImpl& vhost,
...
                                           route.route().host_rewrite_header()))
                                     : absl::nullopt),
+      host_rewrite_path_regex_(
+          route.route().has_host_rewrite_path_regex()
+              ? Regex::Utility::parseRegex(route.route().host_rewrite_path_regex().pattern())
+              : nullptr),
+      host_rewrite_path_regex_substitution_(
+          route.route().has_host_rewrite_path_regex()
+              ? route.route().host_rewrite_path_regex().substitution()
+              : ""),
```
Comment (dio) [inline source/common/router/config_impl.cc:297]: We usually set it as EMPTY_STRING here. But I think we can do cleanup later.

**Pull request merged**